lib/galaxy/objectstore/cloud.py

```diff
@@ -0,0 +1,568 @@
+                    file_list.append(file_tuple)
+            # Sort the file list (based on access time)
+            file_list.sort()
+            # Initiate cleaning once within 10% of the defined cache size?
+            cache_limit = self.cache_size * 0.9
+            if total_size > cache_limit:
+                log.info("Initiating cache cleaning: current cache size: %s; clean until smaller than: %s",
+                         convert_bytes(total_size), convert_bytes(cache_limit))
+                # How much to delete? If simply deleting up to the cache-10% limit,
+                # is likely to be deleting frequently and may run the risk of hitting
+                # the limit - maybe delete additional #%?
+                # For now, delete enough to leave at least 10% of the total cache free
+                delete_this_much = total_size - cache_limit
+                self.__clean_cache(file_list, delete_this_much)
+            self.sleeper.sleep(30)  # Test cache size every 30 seconds?
+
+    def __clean_cache(self, file_list, delete_this_much):
+        """ Keep deleting files from the file_list until the size of the deleted
+        files is greater than the value in delete_this_much parameter.
+
+        :type file_list: list
+        :param file_list: List of candidate files that can be deleted. This method
+            will start deleting files from the beginning of the list so the list
+            should be sorted accordingly. The list must contains 3-element tuples,
+            positioned as follows: position 0 holds file last accessed timestamp
+            (as time.struct_time), position 1 holds file path, and position 2 has
+            file size (e.g., (<access time>, /mnt/data/dataset_1.dat), 472394)
+
+        :type delete_this_much: int
+        :param delete_this_much: Total size of files, in bytes, that should be deleted.
+        """
+        # Keep deleting datasets from file_list until deleted_amount does not
+        # exceed delete_this_much; start deleting from the front of the file list,
+        # which assumes the oldest files come first on the list.
+        deleted_amount = 0
+        for entry in enumerate(file_list):
+            if deleted_amount < delete_this_much:
+                deleted_amount += entry[2]
+                os.remove(entry[1])
+                # Debugging code for printing deleted files' stats
+                # folder, file_name = os.path.split(f[1])
+                # file_date = time.strftime("%m/%d/%y %H:%M:%S", f[0])
+                # log.debug("%s. %-25s %s, size %s (deleted %s/%s)" \
+                #     % (i, file_name, convert_bytes(f[2]), file_date, \
+                #     convert_bytes(deleted_amount), convert_bytes(delete_this_much)))
+            else:
+                log.debug("Cache cleaning done. Total space freed: %s", convert_bytes(deleted_amount))
+                return
+
+    def _get_bucket(self, bucket_name):
```

Comment:
@nuwang; referring to your comment on #4352; the logic here is "get the bucket, if not available, then create it first, then return the bucket". However, the cloudbridge logic is "get the bucket, if not available, return none" ([ see this line](https://github.com/VJalili/cloudbridge/blob/master/cloudbridge/cloud/providers/aws/resources.py#L884)). So, do you want to update cloudbridge and have this logic implemented internally?

Reply:
@VJalili @afgane I think that sequence is alright. I was thinking more about the outer loop - having to loop 5 times before getting a bucket. Calling cloudbridge.create() should guarantee that the bucket is available when it returns, or it should raise an exception. Having to loop suggests that the cloudbridge code is not portable across providers (e.g. this might work in one go on openstack, and a developer would think the code is working correctly, but then things will actually fall over on AWS). We could either document this as being necessary at the interface level, or handle this internally within cloudbridge. Do you know when specifically this occurs?

Reply:
I've created an issue here so we can continue the discussion: https://github.com/gvlproject/cloudbridge/issues/64

Reply:
@nuwang I agree. I removed the retries here, and now we rely on the first try :)

Reply:
@VJalili Ok, sounds good :-) Will close the corresponding issue.